torch/lib/c10d/WinSockUtils.hpp

```diff
@@ -0,0 +1,95 @@
+#pragma once
+
+#include <c10d/Utils.hpp>
+
+namespace c10d {
+namespace tcputil {
+
+#define AF_SELECTED AF_INET
+#define CONNECT_SOCKET_OFFSET 1
+
+inline void closeSocket(int socket) {
+  ::closesocket(socket);
+}
+
+inline int setSocketAddrReUse(int socket) {
+  bool optval = false;
+  return ::setsockopt(
+      socket, SOL_SOCKET, SO_REUSEADDR, (char*)&optval, sizeof(bool));
+}
+
+inline int poll(struct pollfd *fdArray, unsigned long fds, int timeout) {
+  return WSAPoll(fdArray, fds, timeout);
+}
+
+inline void addPollfd(std::vector<struct pollfd>& fds, int socket, short events) {
+  fds.push_back({(SOCKET)socket, events});
+}
+
+inline void waitSocketConnected(
+    int socket,
+    struct ::addrinfo* nextAddr,
+    std::chrono::milliseconds timeout,
+    std::chrono::time_point<std::chrono::high_resolution_clock> startTime) {
+    unsigned long block_mode = 1;
+    SYSCHECK_ERR_RETURN_NEG1(ioctlsocket(socket, FIONBIO, &block_mode));
+
+    int ret;
+    do {
+        ret = connect(socket, nextAddr->ai_addr, nextAddr->ai_addrlen);
+        if (ret == SOCKET_ERROR) {
+            int err = WSAGetLastError();
+            if (err == WSAEISCONN) {
+                break;
+            }
+            else if (err == WSAEALREADY || err == WSAEWOULDBLOCK) {
+            if (timeout != kNoTimeout) {
+                const auto elapsed = std::chrono::high_resolution_clock::now() - startTime;
+                if (elapsed > timeout) {
+                errno = 0;
+                throw std::runtime_error(kConnectTimeoutMsg);
+                }
+            }
+            std::this_thread::sleep_for(std::chrono::milliseconds(10));
+            continue;
+            }
+            throw std::system_error(errno, std::system_category());
+        }
+    } while (ret == SOCKET_ERROR);
+
+    socklen_t errLen = sizeof(errno);
+    errno = 0;
```

Comment:
do we expect non-zero value for `errno` when reaching here? If yes/no, do we need to handle that explicitly?